### PR TITLE
[SPARK-25408] Move to more ideomatic Java8

### PR DESCRIPTION
**Pull request by Fokko**
While working on another PR, I noticed that there is quite some legacy Java in there that can be beautified. For example the use of features from Java8, such as:
- Collection libraries
- Try-with-resource blocks

No logic has been changed. I think it is important to have a solid codebase with examples that will inspire next PR's to follow up on the best practices.

What are your thoughts on this?

This makes code easier to read, and using try-with-resource makes is less likely to forget to close something.

## What changes were proposed in this pull request?

No changes in the logic of Spark, but more in the aesthetics of the code.

## How was this patch tested?

Using the existing unit tests. Since no logic is changed, the existing unit tests should pass.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by cloud-fan** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
mind filling PR description please?

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleIntegrationSuite.java [hunk 1]
```diff
@@ -133,37 +133,38 @@ private FetchResult fetchBlocks(
...
+              if (!res.successBlocks.contains(blockId) && !res.failedBlocks.contains(blockId)) {
+                res.failedBlocks.add(blockId);
+                requestsRemaining.release();
+              }
             }
           }
-        }
-      }, null);
+        }, null
+      );
```
Comment (srowen) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleIntegrationSuite.java]: Nit: move this onto the previous line

File: sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/CLIService.java [hunk 2]
```diff
@@ -154,6 +154,7 @@ public synchronized void start() {
       throw new ServiceException("Unable to connect to MetaStore!", e);
     }
     finally {
+      // IMetaStoreClient is not AutoCloseable, closing it manually
```
Comment (srowen) [inline sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/CLIService.java]: I wouldn't bother with this. This is a copy of some Hive code anyway

**Comment by srowen** [comment]
Jenkins test this please

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleIntegrationSuite.java (hunk 1 above)
Comment (Fokko) [reply common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleIntegrationSuite.java]: Done

**Comment by Fokko** [comment]
Valid points. Personally I'm a fan of explicit final, instead of implicit. But that's a matter of taste :-)

**Comment by cloud-fan** [comment]
retest this please

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolverSuite.java [hunk 3]
```diff
@@ -98,19 +98,19 @@ public void testSortShuffleBlocks() throws IOException {
...
-    InputStream block1Stream =
-      resolver.getBlockData("app0", "exec0", 0, 0, 1).createInputStream();
-    String block1 = CharStreams.toString(
-        new InputStreamReader(block1Stream, StandardCharsets.UTF_8));
-    block1Stream.close();
-    assertEquals(sortBlock1, block1);
+    try (InputStream block0Stream = resolver.getBlockData(
+            "app0", "exec0", 0, 0, 0).createInputStream()) {
+      String block0 =
+              CharStreams.toString(new InputStreamReader(block0Stream, StandardCharsets.UTF_8));
```
Comment (dongjoon-hyun) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolverSuite.java]: Indentation?

File: common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleSecuritySuite.java [hunk 4]
```diff
@@ -96,14 +96,16 @@ private void validate(String appId, String secretKey, boolean encrypt)
...
-      new ExternalShuffleClient(testConf, new TestSecretKeyHolder(appId, secretKey), true, 5000);
-    client.init(appId);
-    // Registration either succeeds or throws an exception.
-    client.registerWithShuffleServer(TestUtils.getLocalHost(), server.getPort(), "exec0",
-      new ExecutorShuffleInfo(new String[0], 0,
-        "org.apache.spark.shuffle.sort.SortShuffleManager"));
-    client.close();
+    try (ExternalShuffleClient client =
+        new ExternalShuffleClient(
+                testConf, new TestSecretKeyHolder(appId, secretKey), true, 5000)) {
```
Comment (dongjoon-hyun) [inline common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/ExternalShuffleSecuritySuite.java]: Indentation? It seems to follow 4-space indentation for Java style, but it seems to be too outstanding in this file.

**Comment by vanzin** [comment]
While we're at it, it would be great to fix the typo in the PR title.

**Comment by Fokko** [comment]
Thanks @dongjoon-hyun. I've fixed the indentation issues.

File: sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/expressions/RowBasedKeyValueBatchSuite.java [hunk 5]
```diff
@@ -152,31 +151,27 @@ public void emptyBatch() throws Exception {
...
-            valueSchema, taskMemoryManager, DEFAULT_CAPACITY);
-    try {
-      Assert.assertEquals(batch1.getClass(), VariableLengthRowBasedKeyValueBatch.class);
-      Assert.assertEquals(batch2.getClass(), FixedLengthRowBasedKeyValueBatch.class);
-    } finally {
-      batch1.close();
-      batch2.close();
+  public void batchType() {
+    try (RowBasedKeyValueBatch batch1 = RowBasedKeyValueBatch.allocate(keySchema,
+        valueSchema, taskMemoryManager, DEFAULT_CAPACITY);
```
Comment (srowen) [inline sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/expressions/RowBasedKeyValueBatchSuite.java]: There's a compile error here. But note you can have multiple Closeables declared in one try block anyway, so these can be merged.

Comment (Fokko) [reply sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/expressions/RowBasedKeyValueBatchSuite.java]: My mistake. thanks!

**Review by srowen — CHANGES_REQUESTED** [review]

File: sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/CLIService.java (hunk 2 above)
Comment (Fokko) [reply sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/CLIService.java]: Done!

**Review by srowen — APPROVED** [review]

**Pull request closed without merging**